org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/hover/LSPTextHover.java

```diff
@@ -193,28 +194,15 @@ public IRegion getHoverRegion(ITextViewer textViewer, int offset) {
 	private void initiateHoverRequest(@NonNull ITextViewer viewer, int offset) {
 		final IDocument document = viewer.getDocument();
 		this.lastViewer = viewer;
-		this.request = LanguageServiceAccessor
-			.getLanguageServers(document, capabilities -> LSPEclipseUtils.hasCapability(capabilities.getHoverProvider()))
-				.thenApplyAsync(languageServers -> // Async is very important here, otherwise the LS Client thread is in
-													// deadlock and doesn't read bytes from LS
-				languageServers.stream()
-					.map(languageServer -> {
-						try {
-								return languageServer.getTextDocumentService()
-										.hover(LSPEclipseUtils.toHoverParams(offset, document))
-										.get(GET_TIMEOUT_MS, TimeUnit.MILLISECONDS);
-						} catch (ExecutionException | BadLocationException e) {
-							LanguageServerPlugin.logError(e);
-							return null;
-						} catch (InterruptedException e) {
-							LanguageServerPlugin.logError(e);
-							Thread.currentThread().interrupt();
-							return null;
-						} catch (TimeoutException e) {
-							LanguageServerPlugin.logWarning("Could not get hover due to timeout after " + GET_TIMEOUT_MS + " miliseconds", e); //$NON-NLS-1$ //$NON-NLS-2$
-							return null;
-						}
-					}).filter(Objects::nonNull).collect(Collectors.toList()));
+		try {
+			HoverParams params = LSPEclipseUtils.toHoverParams(offset, document);
+
+			this.request = LanguageServers.forDocument(document)
+				.withFilter(capabilities -> LSPEclipseUtils.hasCapability(capabilities.getHoverProvider()))
+				.collectAll(server -> server.getTextDocumentService().hover(params));
```

Comment:
Please keep in mind that it should really nice to support cancel checker. For instance with LemMinx Maven, hover can take some time because compute is complex. When user stop the hover, it should cancel all ComplecteableFuture `server.getTextDocumentService().hover(...)`.

This cancel should be done before `this.request = LanguageServers.forDocument(document)`

I spend some times to try to support this cancel by trying to use CompletableFuture API (without success). The only solution that I found is to collect in a list all call of server.getTextDocumentService().hover(..) and loop for the list to cancel it.

See my draft with documentLink but it is the same thing for hover https://github.com/eclipse/lsp4e/pull/238

If you find a better idea with my LSPRequest wrapper, you are welcome!

Reply:
Yes, I agree, supporting a cancel checker is a must, at least for a mature product. Since we are introducing now a new API, I think it makes sense to see it can accommodate the case.

Reply:
Hi @ahmedneilhussain , I think this PR would not prevent handling cancelation as done in #238. One can still wrap the request into something that can be cancelled once a newer request is initiated, right? If so we can merge this one without preventing cancellation handling, and then I would merge this PR independent of the cancellation support which does not exist yet.

In can also refactor it later to add some `withCapability` method to `LanguageServers`